[BugFix] Fix three dict_mapping crash problems (#40897)

srlch opened the pull request:
Problem 1:
In debug/asan mode, dict_mapping function will crash if the given key has multiple column. It is because we forget to resize all key column size before call chunk->update_column_by_index() which make the chunk->check_or_die assert fail in debug/asan.

Problem 2:
If the given key value has null value, BE will crash. dict_mapping does not support null value in key.

Problem 3:
Suppose we have a table with column A as type INT. And we want execute: insert into table (A) values (dict_mapping(...)), BE will crash. This because DictQueryOperator always has return type BIGINT(which is a bug). It will cause optimizer generated a wrong execution plan(miss a necessary cast expression) and make be crash finally.

Solution:
1. construct chunk directly using columns and schema
2. return error if key chunk has null value
3. make sure DictQueryOperator return type is correct.

Fixes #40897

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

trueeyu commented on be/src/exprs/dict_query_expr.cpp:
```diff
@@ -45,6 +45,9 @@ StatusOr<ColumnPtr> DictQueryExpr::evaluate_checked(ExprContext* context, Chunk*
     key_chunk->reset();
+    // make sure all columns in key_chunk have the same size.
+    // Otherwise assert fail in DEBUG/ASAN mode in check_or_die()
+    key_chunk->set_num_rows(size);
```
After key_chunk->reset(), all the size of columns is 0, why check failed?

srlch replied:
In the following, we will call update_column_by_index() to replace the column one by one. Support in the init state, all column is zero size. But after we replace the first column, column_1.size > 0 which is not 0 but it will still call check_or_die() and cause (column_1.size != column_2.size(which is 0)) and fail

trueeyu replied:
Remove the check_or_die()? otherwise the set_num_rows will resize the column, later the column will release, the resize is useless?

srlch replied:
updated

srlch committed:
[BugFix] Fix three dict_mapping crash problems (#40897)

Problem 1:
In debug/asan mode, dict_mapping function will crash if the
given key has multiple column. It is because we forget to
resize all key column size before call chunk->update_column_by_index()
which make the chunk->check_or_die assert fail in debug/asan.

Problem 2:
If the given key value has null value, BE will crash. dict_mapping
does not support null value in key.

Problem 3:
Support we have a table with column A as type INT. And we want execute:
insert into table (A) values (dict_mapping(...)), BE will crash. This because
DictQueryOperator always has return type BIGINT(which is a bug). It will
cause optimizer generated a wrong execution plan(miss a necessary cast expression)
and make be crash finally.

Solution:
1. construct chunk directly using columns and schema
2. return error if key chunk has null value
3. make sure DictQueryOperator return type is correct.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented:
rebase

srlch commented:
rebase

srlch commented:
@Mergify rebase

wyb commented on be/src/exprs/dict_query_expr.cpp:
```diff
@@ -43,17 +43,21 @@ StatusOr<ColumnPtr> DictQueryExpr::evaluate_checked(ExprContext* context, Chunk*
+            return Status::InternalError("invalid parameter : get NULL paramenter");
+        }
         if (column->is_nullable()) {
             column = ColumnHelper::update_column_nullable(false, column, column->size());
```
useless?

srlch replied:
It is not useless. has_null() is different from is_nullable(). The column here can be a nullable column(is_nullable == true) but it must contain all value without any null (has_null() == false)
